src/config.rs

```diff
@@ -311,6 +311,16 @@ impl ConfigBuilder {
             }
         }
 
+        let smtp_username = env::var("SENDGRID_USERNAME");
+        let smtp_password = env::var("SENDGRID_PASSWORD");
+        if smtp_username.is_ok() && smtp_password.is_ok() {
+            println!("SMTP through SendGrid");
+            let smtp_server = "smtp.sendgrid.net:587".to_string();
```

Comment:
This is solution is fine, I think. I'm not sure it's going to be easy to find a single solution we can apply to all providers, as you said yourself.

Though, let's lose the `println!`, and the intermediate `smtp_server` var is also not necessary, fold it into the `self.smtp_server = ...` line.